File: swift/MagCry/Models/Trading.swift
import Foundation

// MARK: - Errors

enum TradingError: Error, LocalizedError {
    case invalidSpread(Int)
    case unknownPlayer(String)
    case selfTrade
    case settlePhase

    var errorDescription: String? {
        switch self {
        case .invalidSpread(let s): return "Spread must be exactly 2, got \(s)"
        case .unknownPlayer(let id): return "Unknown player: \(id)"
        case .selfTrade: return "Cannot trade with yourself"
        case .settlePhase: return "Cannot trade during settlement"
        }
    }
}

// MARK: - Functions

/// Validate that a bid/ask pair has spread of exactly 2.
func validateQuote(bid: Int, ask: Int) throws -> Quote {
    let spread = ask - bid
    guard spread == 2 else { throw TradingError.invalidSpread(spread) }
    return Quote(bid: bid, ask: ask)
}

/// Execute a trade directly at a specified price (V2 mechanic).
/// This is the primary trade function — no standing quote needed.
@discardableResult
func executeTradeDirectly(
    state: GameState,
    buyer: String,
    seller: String,
    price: Int
) throws -> Trade {
    guard buyer != seller else { throw TradingError.selfTrade }
    guard state.phase != .settle else { throw TradingError.settlePhase }
    guard state.playerIDs.contains(buyer) else { throw TradingError.unknownPlayer(buyer) }
    guard state.playerIDs.contains(seller) else { throw TradingError.unknownPlayer(seller) }

    let trade = Trade(buyer: buyer, seller: seller, price: price, phase: state.phase)
    state.trades.append(trade)
    return trade
}

File: kotlin/app/src/main/java/com/magcry/model/Trading.kt
package com.magcry.model

// -- Errors --

sealed class TradingException(message: String) : Exception(message) {
    class InvalidSpread(spread: Int) : TradingException("Spread must be exactly 2, got $spread")
    class UnknownPlayer(id: String) : TradingException("Unknown player: $id")
    class SelfTrade : TradingException("Cannot trade with yourself")
    class SettlePhase : TradingException("Cannot trade during settlement")
}

// -- Functions --

/** Validate that a bid/ask pair has spread of exactly 2. */
fun validateQuote(bid: Int, ask: Int): Quote {
    val spread = ask - bid
    if (spread != 2) throw TradingException.InvalidSpread(spread)
    return Quote(bid = bid, ask = ask)
}

/** Execute a trade directly at a specified price (V2 mechanic). */
fun executeTradeDirectly(
    state: GameState,
    buyer: String,
    seller: String,
    price: Int
): Trade {
    if (buyer == seller) throw TradingException.SelfTrade()
    if (state.phase == Phase.SETTLE) throw TradingException.SettlePhase()
    if (buyer !in state.playerIDs) throw TradingException.UnknownPlayer(buyer)
    if (seller !in state.playerIDs) throw TradingException.UnknownPlayer(seller)

    val trade = Trade(buyer = buyer, seller = seller, price = price, phase = state.phase)
    state.trades.add(trade)
    return trade
}
